### DISCUSSION TITLE
Custom counter for syslog messages

### DISCUSSION THREAD

**Discussion by M-a-x-Z** [syslog-ng/syslog-ng#4721] [Q&A] [answered]
Is there any way to obtain sequently numbers for every processing message? Someting like RCPTID not for entire server but for every destination?

Maybe there is any possibility to increment global variables in rewrite block or there is way to use some built-in counters?

**Comment by bazsi** [comment]
$SEQNUM is a per-destination counter, but is only applied to local messages and not forwarded ones. It is the implementation for RFC5424  meta.sequenceId structured data field. If you can explain your use-case in more details I'd be willing to implement something like this. Where do you send these messages, how would you process the counter?

thanks

**Comment by M-a-x-Z** [comment]
In my monitoring system, syslog-ng receives messages via syslog udp from many network devices, then moves them to the pipe and passes them to a python script. It is necessary to know: has each massage been analyzed or the pipe been overloaded? It seems to me that the best way is to number the messages placed in the pipe. Syslog-ng has an RCPTID that knows how to number messages, but does it globally for the entire server. That is, in my case, the numbers come with omissions: 456, 489, 503, etc.

I want to add a counter that will count the received messages separately for each destination, starting from 1 from the moment the daemon starts. So i will know, if any message will be lost due to python script overload.

I think I've found a workaround:

`
python {
    class TC_CNT(object):
        def init(self):
            self.counter = 0
            return True
        def parse(self, log_message):
            self.counter += 1
            log_message['CNT'] = str(self.counter)
            return True
};

parser my_python_parse { python(class("TC_CNT")); }
`
but i need update production (for python support) to test it. It is not easy. because it needs to update ubuntu release.
What do you think about this solution?

**Reply by bazsi** [reply]
This can potentially work, however using Python for this purpose may limit syslog-ng's performance to be less than 50k msg/sec, which might be fine for you (just a guess, you'd have to measure this if this can be a problem).

On the other hand, if it's syslog-ng that handles over messages to your Python script, you can always query the dropped counter for every destination. That should tell you if syslog-ng dropped messages because it was unable to send it onwards.

You can query the internal counters for syslog-ng using the "syslog-ng-ctl" stats interface. For instance, I have am unnamed tcp destination  to port 2001 and I have its dropped counter here:

```
$ /install/sbin/syslog-ng-ctl stats | grep dropped
dst.tcp;#anon-destination1#0;tcp,127.0.0.1:2001;a;dropped;0
```

But with newer versions of syslog-ng (4.2+ I think), you can also get the same information in a format similar to what prometheus uses:

```
$ /install/sbin/syslog-ng-ctl stats prometheus | grep dropped
syslogng_output_events_total{transport="tcp",address="127.0.0.1:2001",driver="afsocket",id="#anon-destination1#0",result="dropped"} 0
```

syslog-ng would only drop messages if our local buffer becomes full (either memory or disk based). I understand that the sequence number would allow you to detect the same drops on the server side, without collecting syslog-ng drop counters, but actually I think this is a better solution long term, especially as syslog-ng has other interesting metrics too.

**Comment by bazsi** [comment]
I've just opened a PR that would help you: https://github.com/syslog-ng/syslog-ng/pull/4745 

if you add flags(seqnum-all) to your destination, you'll get a $SEQNUM value that gets incremented for every message. this departs from the behaviour specified in RFC5424 but that did not make much sense anyway. The default is to keep rfc5424 behaviour.

**Comment by bazsi** [comment]
btw and you could use AxoSyslog to try new features without upgrading the OS. That ships syslog-ng in a container.
Here's the github page: https://github.com/axoflow/axosyslog-docker

And the documentation: https://axoflow.com/docs/axosyslog-core/

**Comment by M-a-x-Z** [comment]
> This can potentially work, however using Python for this purpose may limit syslog-ng's performance to be less than 50k msg/sec, which might be fine for you (just a guess, you'd have to measure this if this can be a problem).

Thanks a lot, 50k msg/sec will be quite enough

> On the other hand, if it's syslog-ng that handles over messages to your Python script, you can always query the dropped counter for every destination.

I have no doubt about syslog-ng performance. However i put the data from syslog-ng into the PIPE and it may be droped there if the PIPE overflows before it is read. It's not going to consider into syslog-ng counters, is it?

> if you add flags(seqnum-all) to your destination, you'll get a $SEQNUM value that gets incremented for every message. this departs from the behaviour specified in RFC5424 but that did not make much sense anyway. The default is to keep rfc5424 behaviour.

do I understand correctly that without 'seqnum-all', the seqnum is taken from the content of the message datagram? If many switches send messages to the same port, will each have its own sequence for seqnum?

**Comment by bazsi** [comment, marked as answer]
> > This can potentially work, however using Python for this purpose may limit syslog-ng's performance to be less than 50k msg/sec, which might be fine for you (just a guess, you'd have to measure this if this can be a problem).
> 
> Thanks a lot, 50k msg/sec will be quite enough
> 
> > On the other hand, if it's syslog-ng that handles over messages to your Python script, you can always query the dropped counter for every destination.
> 
> I have no doubt about syslog-ng performance. However i put the data from syslog-ng into the PIPE and it may be droped there if the PIPE overflows before it is read. It's not going to consider into syslog-ng counters, is it?

syslog-ng will properly implement flow control for pipes, e.g. will not write them unless it can be delivered. So pipe() destinations are reliable.


> 
> > if you add flags(seqnum-all) to your destination, you'll get a $SEQNUM value that gets incremented for every message. this departs from the behaviour specified in RFC5424 but that did not make much sense anyway. The default is to keep rfc5424 behaviour.
> 
> do I understand correctly that without 'seqnum-all', the seqnum is taken from the content of the message datagram? If many switches send messages to the same port, will each have its own sequence for seqnum?


It will be taken from the original message yes, at least if the client sent one.

If they did send their own seqnum, then all of them will have a separate counter.

But with seqnum-all syslog-ng will have continuously increasing seqnum on output, that will override the one we received.

**Reply by M-a-x-Z** [reply]
> syslog-ng will properly implement flow control for pipes, e.g. will not write them unless it can be delivered. So pipe() destinations are reliable.

Thank you, so I can actually use syslog-ng dropped packets counter.

> But with seqnum-all syslog-ng will have continuously increasing seqnum on output, that will override the one we received.

Yes, this is another solution of my problem
